src/rpm/systemd-update-helper.c

```diff
@@ -0,0 +1,912 @@
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "%s expects at least a single unit argument", argv[0]);
+
+                STRV_FOREACH(unit, units)
+                        if (!unit_name_is_valid(*unit, UNIT_NAME_ANY))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "%s is not a valid unit name", *unit);
+        }
+
+        *ret = TAKE_PTR(units);
+        return !strv_isempty(*ret);
+}
+
+static int expand_template_units(sd_bus *bus, char **units, char ***ret) {
+        _cleanup_strv_free_ char **sv = NULL, **globs = NULL, **expanded = NULL;
+        int r;
+
+        assert(ret);
+
+        STRV_FOREACH(unit, units) {
+                UnitNameFlags flags = unit_name_classify(*unit);
+                if (flags == -EINVAL) {
+                        log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "%s is not a valid unit name, ignoring", *unit);
+                        continue;
+                }
+                if (flags < 0)
+                        return flags;
+
+                if (flags & UNIT_NAME_TEMPLATE) {
+                        _cleanup_free_ char *glob = NULL;
+
+                        r = unit_name_replace_instance_full(*unit, "*", /* accept_glob= */ true, &glob);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to turn template unit into glob: %m");
+
+                        if (strv_consume(&globs, TAKE_PTR(glob)) < 0)
+                                return log_oom();
+                } else
+                        if (strv_extend(&sv, *unit) < 0)
+                                return log_oom();
+        }
+
+        r = list_units(bus, globs, &expanded);
+        if (r < 0)
+                return r;
+
+        if (strv_extend_strv_consume(&sv, TAKE_PTR(expanded), /* filter_duplicates= */ true) < 0)
+                return log_oom();
+
+        *ret = TAKE_PTR(sv);
+        return 0;
+}
+
+static bool offline(void) {
+        if (running_in_chroot_or_offline())
+                return true;
+
+        if (sd_booted() <= 0)
+                return true;
+
+        return false;
+}
+
+static int bus_connect_user_unit(const char *unit, sd_bus **ret) {
+        int r;
+
+        assert(unit);
+        assert(ret);
+
+        _cleanup_free_ char *user = NULL;
+        r = unit_name_to_instance(unit, &user);
+        if (r < 0)
+                return log_error_errno(r, "Failed to extract user id from unit '%s': %m", unit);
+
+        _cleanup_free_ char *host = strjoin(user, "@");
+        if (!host)
+                return log_oom();
+
+        r = bus_connect_transport(BUS_TRANSPORT_MACHINE, host, RUNTIME_SCOPE_USER, ret);
+        if (r < 0)
+                return bus_log_connect_full(arg_quiet ? LOG_DEBUG : LOG_WARNING, r,
+                                            BUS_TRANSPORT_MACHINE, RUNTIME_SCOPE_USER);
```

Comment:
gah, i really head connecting to user context from system context like this, and even in a blocking fashion...